### ISSUE TITLE
tag a release?

### ISSUE THREAD

**Issue by vbatts** [rootless-containers/slirp4netns#44]
Even if just an 0.1 for landmarking as a packager

**Comment by AkihiroSuda** [comment]
As Fedora pkg is already versioned with `0.1-%d.dev.git%s`, maybe better to start with v0.2?


cc package maintainers @lsm5 @ansemjo @vrothberg (who is maintaining SUSE pkg now?) Any thought?

**Comment by ansemjo** [comment]
The [slirp4netns-git](https://aur.archlinux.org/packages/slirp4netns-git/) package is simply using the number of commits as a version number (i.e. mine is currently at `r62.777bdcc-1`) as is described in the [VCS package guidelines](https://wiki.archlinux.org/index.php/VCS_package_guidelines#Git) because there were no tags to use.

If you were to start using tags, starting with `0.2` or `0.1.1` would make sense to me. I'd probably create a second package for the tagged releases then so that you have both `slirp4netns` and `slirp4netns-git` in the AUR.

**Comment by ansemjo** [comment]
It would require build changes though because the version `0.1` is currently hardcoded:
https://github.com/rootless-containers/slirp4netns/blob/5e4789b9a21cab3315801e1750670e10a1c86a2a/configure.ac#L2

**Comment by AkihiroSuda** [comment]
Is it possible to generate the AC_INIT version string from git tag?

**Comment by ansemjo** [comment]
I'm testing something to try and define `VERSION` with `-DVERSION="\"$(VERSION)\""`, similarly to how I did that for `COMMIT`.

**Comment by ansemjo** [comment]
See the pull request. After merging it you could start tagging commits in git and it will be built into the binary.

I wrote [ansemjo/version.sh](https://github.com/ansemjo/version.sh) specifically after my last pull request in this repository to have one *single source of truth* for version information usable in various languages.

**Comment by AkihiroSuda** [comment]
Thanks!

**Comment by AkihiroSuda** [comment]
Also we need to consider git branch management.
I suppose we can just use single `master` branch for now.
In future we can create `release/0.2` branch (and we should consider its EOL policy), after we add major feature for v0.3+ && want to backport bug-fix to v0.2.

**Comment by giuseppe** [comment]
@ansemjo @AkihiroSuda could we use https://github.com/coreutils/gnulib/blob/master/build-aux/git-version-gen ?

It is used by different GNU programs to generate the version string using git tags.

For example we can tag something like v0.1 on git and the `./configure` will automatically generate the correct version used by the program.

**Comment by vrothberg** [comment]
> As Fedora pkg is already versioned with `0.1-%d.dev.git%s`, maybe better to start with v0.2?
> 
> cc package maintainers @lsm5 @ansemjo @vrothberg (who is maintaining SUSE pkg now?) Any thought?

@cyphar, @flavio and @marcov will know.

**Comment by cyphar** [comment]
I think starting with v0.2 would be a good idea if existing packages use a fake version.

On the topic of branch maintenance, I would recommend not over-complicating maintenance until it's necessary. In particular, just doing commits from master for each new release will work okay for a project like this (until we have to start making breaking changes).

As for versioning with autoconf, [in catatonit](https://github.com/openSUSE/catatonit) I have the version in `configure.ac` and then generate `VERSION` using `VERSION.in` (with `@PROGRAM_VERSION@`). In addition, `PROGRAM_VERSION` is defined in `config.h`. This has worked okay for me.

**Comment by ansemjo** [comment]
> @ansemjo @AkihiroSuda could we use https://github.com/coreutils/gnulib/blob/master/build-aux/git-version-gen ?

I suppose so. It's not clear to me when the `.version` file is generated though or if it works from where no `.git` is present, as @cyphar says.

> ... and I would strongly suggest against getting the version from .git during builds. In OBS we don't have a .git tree and so the versions would all be "unknown".

I am not familiar with OBS. Is that because it uses tarballs of the repository instead of clones? Because that's what [version.sh](https://github.com/ansemjo/version.sh) is supposed to solve by using `.gitattributes`.

**Comment by cyphar** [comment]
> I am not familiar with OBS. Is that because it uses tarballs of the repository instead of clones?

Right, we use release tarballs which generally don't contain `.git` (but also given that we might apply patches and other such things, having a `VERSION` file comitted that is used for the version string is infinitely better than using the `git tag` for the build).

Speaking of which I would recommend using the script that `runc` and `umoci` use (which GPG-signs the assets) for making release assets.

**Comment by ansemjo** [comment]
So, to sum up: #45 would enable embedding a version in the binary, similarly to how it's done for the commit hash already. But you'd still need some sort of release process to e.g. create signed release assets and decide when to tag a new version? The script @cyphar refers to for this purpose is this one: https://github.com/openSUSE/umoci/blob/v0.4.3/hack/release.sh

**Comment by AkihiroSuda** [comment]
@vbatts WDYT?

**Comment by vbatts** [comment]
Having a fit tag in the version is clunky. It seems that having a config.h that is revved when you make a git tag makes most sense.

**Comment by AkihiroSuda** [comment]
> It seems that having a config.h that is revved when you make a git tag makes most sense.

Just updating `configure.ac` manually is ok for now? (as in fuse-overlayfs)
https://github.com/rootless-containers/slirp4netns/blob/5e4789b9a21cab3315801e1750670e10a1c86a2a/configure.ac#L2

**Comment by vbatts** [comment]
Also good

-------- Original Message --------

**Comment by AkihiroSuda** [comment]
PR: https://github.com/rootless-containers/slirp4netns/pull/51
